vllm/distributed/eplb/policy/default.py

```diff
@@ -207,6 +207,110 @@ def inverse(perm: torch.Tensor) -> torch.Tensor:
         logcnt = mlogcnt.view(num_layers, -1).gather(-1, log2mlog)
         return pphy2log, pphyrank, logcnt
 
+    @classmethod
+    def preserve_intragpu_slots(
+        cls,
+        phy2log: torch.Tensor,
+        phyrank: torch.Tensor,
+        num_ranks: int,
+        old_global_expert_indices: torch.Tensor,
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        """
+        Reorder the new mapping per GPU so that experts that remain on the same GPU
+        keep their previous slot positions when possible. Incoming experts to that GPU
+        fill any remaining available slots. This is applied only when the number of GPUs
+        is unchanged and the slots per GPU remain the same between
+        the old and new mappings.
+        """
+        device = phy2log.device
+        new_num_phy = phy2log.shape[1]
+        old_num_phy = old_global_expert_indices.shape[1]
+        if (
+            num_ranks <= 0
+            or new_num_phy % num_ranks != 0
+            or old_num_phy % num_ranks != 0
+            or (new_num_phy // num_ranks) != (old_num_phy // num_ranks)
+        ):
+            return phy2log, phyrank
+
+        # Move to CPU and convert to NumPy for processing
+        phy2log_np = phy2log.cpu().numpy()
+        phyrank_np = phyrank.cpu().numpy()
+        old_np = old_global_expert_indices.cpu().numpy()
+
+        slots_per_gpu = new_num_phy // num_ranks
+        num_layers = phy2log_np.shape[0]
+
+        post_phy2log_np = phy2log_np.copy()
+        post_phyrank_np = phyrank_np.copy()
+
+        for gpu_idx in range(num_ranks):
+            start = gpu_idx * slots_per_gpu
+            end = start + slots_per_gpu
+            # Segments across all layers for this GPU
```

Comment:
I think that introducing the term `segment` is a bit confusing here. I think the prefix `local_` on the per-gpu slices would be a bit more readable? Let's also avoid shortening the dimension names in the comments.